node/network/statement-distribution/src/lib.rs

```diff
@@ -911,38 +1098,249 @@ async fn report_peer(
-async fn handle_incoming_message<'a>(
+async fn handle_incoming_message_and_circulate<'a>(
 	peer: PeerId,
-	peer_data: &mut PeerData,
-	active_heads: &'a mut HashMap<Hash, ActiveHeadData>,
-	ctx: &mut impl SubsystemContext<Message = StatementDistributionMessage>,
+	peers: &mut HashMap<PeerId, PeerData>,
 	message: protocol_v1::StatementDistributionMessage,
+	active_heads: &'a mut HashMap<Hash, ActiveHeadData>,
+	ctx: &mut impl SubsystemContext,
+	req_sender: &mpsc::Sender<RequesterMessage>,
 	metrics: &Metrics,
-) -> Option<(Hash, &'a StoredStatement)> {
-	let (relay_parent, statement) = match message {
-		protocol_v1::StatementDistributionMessage::Statement(r, s) => (r, s),
-	};
+) {
+	let relay_parent = message.get_metadata().relay_parent;
+	let result = retrieve_statement_from_message(
+		peer,
+		message,
+		active_heads,
+		ctx,
+		req_sender,
+	).await;
+
+	if let Some((active_head, statement)) = result {
+		handle_incoming_statement_and_circulate(
+			peer,
+			peers,
+			active_head,
+			ctx,
+			relay_parent,
+			statement,
+			metrics,
+		).await;
+	}
+}
 
-	let active_head = match active_heads.get_mut(&relay_parent) {
+/// If message contains a statement, then retrieve it, otherwise fork task to fetch it.
+///
+/// This function will also return `None` if the message did not pass some basic checks, in that
+/// case no statement will be requested, on the flipside you get `ActiveHeadData` in addition to
+/// your statement.
+///
+/// If the message was large, but the result has been fetched already that one is returned.
+async fn retrieve_statement_from_message<'a>(
+	peer: PeerId,
+	message: protocol_v1::StatementDistributionMessage,
+	active_heads: &'a mut HashMap<Hash, ActiveHeadData>,
+	ctx: &mut impl SubsystemContext,
+	req_sender: &mpsc::Sender<RequesterMessage>,
+) -> Option<(&'a mut ActiveHeadData, SignedFullStatement)> {
+
+	let metadata = message.get_metadata();
+
+	let active_head = match active_heads.get_mut(&metadata.relay_parent) {
 		Some(h) => h,
 		None => {
 			tracing::debug!(
 				target: LOG_TARGET,
-				?peer,
-				?relay_parent,
-				"Unknown (or outdated) relay parent"
+				requested_relay_parent = %metadata.relay_parent,
+				"our view out-of-sync with active heads; head not found",
 			);
 			report_peer(ctx, peer, COST_UNEXPECTED_STATEMENT).await;
-			return None;
+			return None
+		}
+	};
+
+	match message {
+		protocol_v1::StatementDistributionMessage::Statement(_, statement) =>
+			return Some((active_head, statement)),
+
+		protocol_v1::StatementDistributionMessage::LargeStatement(metadata) => {
+
+			let candidate_hash = metadata.candidate_hash;
+
+			match active_head.waiting_large_statements.entry(candidate_hash) {
+				Entry::Occupied(mut occupied) => {
+
+					let mut new_status = None;
+
+					match occupied.get_mut() {
+						LargeStatementStatus::Fetching(info) => {
+							info.available_peers.insert(peer, metadata);
+							info.peers_to_try.push(peer);
```

Comment:
Is duplicate protected somewhere before here?

Reply:
Good point, I rely on the duplicate detection later on, but adding duplicate peers to `peers_to_try` can be avoided right away. Is this what you meant?